File: static/nothree.ts
import {Decimal} from "decimal.js";


function colinear(p1: [number, number], 
                  p2: [number, number],
                  p3: [number, number]): boolean {
  // From https://math.stackexchange.com/a/405981/15140
  const [a, b] = p1;
  const [m, n] = p2;
  const [x, y] = p3;
  return (n - b) * (x - m) == (y - n) * (m - a);
}


function couple(lst: number[]): [number, number][] {
  const result: [number, number][] = [];
  for (let i = 0; i < lst.length - 1; i += 2) {
    result.push([lst[i], lst[i+1]]);
  }
  return result;
}


export async function score(nums: Decimal[]): Promise<number | string> {
  const ns = nums.map((num) => num.toNumber());
  for (const num of ns) {
    if (num < 0 || num >= 100) {
      return "Coordinate value not in range [0, 99]!";
    }
  }
  if (ns.length % 2 == 1) {
    return "Not a CSV of points! (Need even length CSV!)";
  }
  const points: [number, number][] = couple(ns);
  if (points.length > 200) {
    return "Too many points!";
  }
  const dupCheck = new Set<[number, number]>(points);
  if (dupCheck.size != points.length) {
    return "All points should be distinct!";
  }
  for (let i = 0; i < points.length; i++) {
    for (let j = i+1; j < points.length; j++) {
      for (let k = j+1; k < points.length; k++) {
        if (colinear(points[i], points[j], points[k])) {
          return "Invalid configuration! 3 points are colinear!";
	}
      }
    }
  }
  return (1 - (points.length / 200)) * 10 ** 6;
}

File: app/routers/problems/nothree.py
from itertools import batched, combinations


def colinear(p1, p2, p3):
    # From https://math.stackexchange.com/a/405981/15140
    a, b = p1
    m, n = p2
    x, y = p3
    return (n - b) * (x - m) == (y - n) * (m - a)


async def score(nums: list[int]):
    # All coordinate values must be in range [0, 99]
    for num in nums:
        if num < 0 or num >= 100:
            return None
    # Need two values for each point, so length should be even.
    if len(nums) % 2 == 1:
        return None
    points = list(batched(nums, 2))
    dups = set(points)
    # All points should be distinct!
    if len(dups) != len(points):
        return None
    # Cannot have more than 2*n (n = 100)
    if len(points) > 200:
        return None

    # Now check if any of the points are colinear, if so, not a valid config.
    for p1, p2, p3 in combinations(points, 3):
        if colinear(p1, p2, p3):
            return None

    return (200 - len(points)) * 5000
